### ISSUE TITLE
converting start/end position in terms of bar

### ISSUE THREAD

**Issue by shinysup** [craffel/pretty-midi#119]
Is there a way represent the start/end position of notes in terms of bar, and vice versa?
(I don't know if "bar" is the right term. I mean the a musical length period that is equal to 4 successive notes of quarter lengths.)

**Comment by craffel** [comment]
Not really.  The closest you can currently get is using `time_to_tick`, which will convert time in seconds to tick, which is tangentially related to metrical position.  Alternatively, a simple hack would be to call `get_beats` and `get_downbeats`, and then assemble the bar/beat position according to the returned values.

I have sometimes thought about making a function which converts from absolute time in seconds to bar/beat position, and vice versa.  It was mostly held up by the fact that there was no way to change both the wall-clock and metrical timing of a `PrettyMIDI` object, but now that has been fixed in #28/#70.  I will leave this issue open as a reminder to implement something like this.

**Comment by shinysup** [comment]
Thanx for the reply. 

Given the answer, is there a way to check whether a track is monophonic? Or, should I check for notes having the same start position?

**Comment by craffel** [comment]
By monophonic hear I assume you mean that only at most 1 note is playing at a given point in time (i.e. no chords)?  A simple and inefficient way to do that would be
```Python
# instrument can be retrieved from the PrettyMIDI.instruments list
monophonic = True
for note1 in instrument.notes:
    for note2 in instrument.notes:
        if note1.start > note2.start and note1.start < note2.end:
            monophonic = False
```

**Comment by bzamecnik** [comment]
Test for an instrument track being monophonic - probably not the most efficient way:

```python
def is_mono(instrument):
    return (instrument.get_piano_roll() > 0).sum(axis=0).max() <= 1
```

Note it depends on `get_piano_roll()` `fs` argument (sampling frequency).

**Comment by bzamecnik** [comment]
As for obtaining the position within the metric grid (measure, beat) actually the MIDI may not contain enough information to precisely determine that. Consider the case with smoothly changing tempo represented just by varying note lengths in ticks, not by tempo change messages. We would have to estimate the tempo curve (instantaneous tempo function + phase of beats and measures) and then perform a mapping to the space of measures. In addition there might be changes in the key signature (eg. 4/4 -> 3/4), so the number of beats per measure may change as well.

**Comment by craffel** [comment]
> Consider the case with smoothly changing tempo represented just by varying note lengths in ticks, not by tempo change messages. 

I actually have found this to be pretty rare in MIDI files found "in the wild", but it does happen sometimes.  That's what the `estimate_tempi` function is for, at least in part.  Estimating the metrical structure is extremely nontrivial though (at least it's easier than for audio, ostensibly :) )... so not within the scope of `pretty_midi`, I don't think.

**Comment by soorejmg** [comment]
> Not really. The closest you can currently get is using `time_to_tick`, which will convert time in seconds to tick, which is tangentially related to metrical position. Alternatively, a simple hack would be to call `get_beats` and `get_downbeats`, and then assemble the bar/beat position according to the returned values.
> 
> I have sometimes thought about making a function which converts from absolute time in seconds to bar/beat position, and vice versa. It was mostly held up by the fact that there was no way to change both the wall-clock and metrical timing of a `PrettyMIDI` object, but now that has been fixed in #28/#70. I will leave this issue open as a reminder to implement something like this.

Hi Craffel, is this feature curently available? Could please give a hint on this?

Thanks
Soorej

**Comment by soorejmg** [comment]
Hi Craffel, 
Basically, I would like to retrieve all time signature bars from a midi file. Is it possible? Please advice.

Thanks
Soorej

**Comment by craffel** [comment]
> Hi Craffel, is this feature curently available?

No, it's not a big priority for me, but a pull request would be welcome.

> I would like to retrieve all time signature bars from a midi file.

You can use [`get_downbeats`](https://github.com/craffel/pretty-midi/blob/master/pretty_midi/pretty_midi.py#L677) which is not too far off from what (I think) you're asking for.

**Comment by soorejmg** [comment]
> > Hi Craffel, is this feature curently available?
> 
> No, it's not a big priority for me, but a pull request would be welcome.
> 
> > I would like to retrieve all time signature bars from a midi file.
> 
> You can use [`get_downbeats`](https://github.com/craffel/pretty-midi/blob/master/pretty_midi/pretty_midi.py#L677) which is not too far off from what (I think) you're asking for.

HI Craffel,
Thanks for this. let me have a try,

Thanks
Soorej

**Comment by soorejmg** [comment]
hi Craffel,
![image](https://user-images.githubusercontent.com/47497915/85985975-160da680-ba1e-11ea-96cb-f37404f30d8f.png)

In this screenshot, is there a way to know whether each note is a quarter note or full note or half note and so on? Any specific method will help?

Thanks
Soorej

**Comment by craffel** [comment]
No, `pretty_midi` does not support this.

**Comment by soorejmg** [comment]
Hi Craffel, 

Could you help in this. Why is there a estimate_beat_start when there is already a get_beats which returns all the beat location in seconds. A bit unclear, not sure if i am missing something. Please advise

![image](https://user-images.githubusercontent.com/47497915/86198735-db754c80-bb8a-11ea-8203-14afcb9c6db4.png)
Thanks
Soorej

**Comment by soorejmg** [comment]
Hi Craffel, 

For the Piano roll function, it is not clear exactly is the functioality. When i tried, always getting zero value as below. Could you see please.

![image](https://user-images.githubusercontent.com/47497915/86201536-686fd400-bb92-11ea-9bbd-a46d3132acf2.png)


Thanks
Soorej

**Comment by craffel** [comment]
> Could you help in this. Why is there a estimate_beat_start when there is already a get_beats which returns all the beat location in seconds. A bit unclear, not sure if i am missing something. Please advise

`estimate_beat_start` is for MIDI files that have corrupt/missing timing information. You should not need this function for the vast majority of MIDI files.

> For the Piano roll function, it is not clear exactly is the functioality. When i tried, always getting zero value as below. Could you see please.

Is `midi_data` empty? It will produce an empty piano roll for an empty `PrettyMIDI` object. You can try the example.mid file in the repository root instead.

**Comment by soorejmg** [comment]
> > Could you help in this. Why is there a estimate_beat_start when there is already a get_beats which returns all the beat location in seconds. A bit unclear, not sure if i am missing something. Please advise
> 
> `estimate_beat_start` is for MIDI files that have corrupt/missing timing information. You should not need this function for the vast majority of MIDI files.
> 
> > For the Piano roll function, it is not clear exactly is the functioality. When i tried, always getting zero value as below. Could you see please.
> 
> Is `midi_data` empty? It will produce an empty piano roll for an empty `PrettyMIDI` object. You can try the example.mid file in the repository root instead.

Hi Craffel,

The midi files where not empty, pfa the ones that i tried with. Just changed the extension as jpg to upload. You could download and rename as midi.

![Anger_Abm_145bpm_7 - Copy](https://user-images.githubusercontent.com/47497915/86300277-f1d7e280-bc34-11ea-91b4-5cc7d4484407.jpg)
![Anger - Copy](https://user-images.githubusercontent.com/47497915/86300302-03b98580-bc35-11ea-836e-0413b23df80b.jpg)

Thanks
Soorej

**Comment by craffel** [comment]
The files you uploaded do not produce empty piano rolls.

```Python
import pretty_midi
pm = pretty_midi.PrettyMIDI("Anger_Abm_145bpm_7 - Copy)
import numpy as np
np.all(pm.get_piano_roll() == 0)
np.nonzero(pm.get_piano_roll())
```

I'm sorry but I can't continue to provide this kind of support - these GitHub issues are for reporting bugs, not for helping with usage of the library.

**Comment by soorejmg** [comment]
Hi Craffel,

Thank you so much for these inputs and our sincere apologies for all those silly questions. Just started using the library and getting used with the terminologies and methods. As  the domain is music which has multiple terms for the same thing, finding it a bit difficult to understand whether the intentions of certain methods are what we expect to be. Will  surely limit our questions are will post everything together. 

Thanks
Soorej

**Comment by Mannerow** [comment]
> Test for an instrument track being monophonic - probably not the most efficient way:
> 
> ```python
> def is_mono(instrument):
>     return (instrument.get_piano_roll() > 0).sum(axis=0).max() <= 1
> ```
> 
> Note it depends on `get_piano_roll()` `fs` argument (sampling frequency).

Hello. I tried this and it works pretty well for determining mono instruments. Can you shed some light on why this works?

Thanks
